###10.1 常见的错误和异常  
1. 拼写错误  
2. 单引号和双引号的混乱  
3. 括号使用混乱  
4. 等号与赋值混淆
5. 缺少美元符号  
6. 调用不存在的常量和变量  
7. 调用不存在的文件  
8. 环境配置的错误  
9. 数据库服务器连接错误  
###10.2错误处理  
####10.2.1 php中的错误处理机制
|  名称     |    默认值        |           含义                   |
|:---:|-----|:---:|
|display_errors|On|设置错误作为  PHP 的一部分输出。开发的过程中可以采用默认的设置，但是为了安全考虑，在生产环境中还是设置为 Off 比较好|
|error_reporting|E_all|这个设置会显示所有的出错信息。这种设置会让一些无措的提示也显示，所以可以设置  error_reportiong 的默认值： error_reporting - E_ALL & ~E_NOTICE ,这样只会显示错误和不良编码|
|error_log|null|设置记录错误日志的文件。默认情况下将错误发送到  Web 服务器日志，用户也可以制定写入的文件|
|html_errors|On|控制是否在错误信息中采用  HTML 格式|
|log_errors|Off|控制是否应该将错误发送到主机服务器的日志文件|
|display_startup_errors|Off|控制是否显示  PHP 启动时的错误|
|track_errors|Off|设置是否保存最近一个警告或错误信息|
####10.2.2 实例1————应用 DIE 语句调试
使用 DIE 语句调试的优势是，不仅可以显示错误的位置，还可以输出错误信息。一旦出现错误，程序将会终止运行，并在浏览器上显示出错之前的信息和错误信息。  
if(!file_exists("welcome.txt")){  
die("文件不存在");  
}else{  
$file=fopen("welcome.txt","r");  
}  
####10.2.3 实例2————自定义错误和错误触发器
自定义错误函数的语法格式：  
error_function(error_level,error_message,error_file,error_line,error_context)  
该函数必须至少包含 error_level 和 error_message 两个参数，另外 3 个参数   error_file、error_line 和   error_context 是可选的。各个参数具体含义如下：  
|参数|含义|
|:---:|:---:|
|error_level|必须参数。为用户定义的错误规定错误报告级别。必须是一个数值|
|error_message|必须参数。为用户定义的错误规定错误消息|
|error_file|可选参数。规定错误在其中发生的文件名|
|error_line|可选参数。规定错误发生的行号|
|error_context|可选参数。规定一个数组，包含了当错误发生时在用的每个变量以及它们的值|

参数 error_level 为定义错误规定的报告级别，这些错误报告级别是错误处理程序处理错误的不同类型。具体的级别值和含义如表：  
|数值|常量|含义|
|:---:|-----|:---:|
|2|E_WARNINO|非致命 run-time 错误。不暂停脚本执行|
|8|E_NOTICE|run-time 通知。脚本发现可能有错误发生，但也可能在脚本征程运行时发生|
|256|E_USER_ERROR|致命的用户生成的错误。这类似于程序员使用  PHP 函数  trigger_error()设置的 E_ERROR|
|512|E_USER_WARNING|非致命的用户生成的警告。这类似于程序员使用  PHP 函数 trigger_error() 设置的  E_WARNING |
|1024|E_USER_NOTICE|用户生成的通知。这类似于程序员使用  PHP 函数 trigger_error() 设置的 E_NOTICE|
|4096|E_RECOVERABLE_ERROR|可捕获的致命错误。类似于  E_ERROR,但可被用户定义的处理程序捕获|
|8191|E_ALL|所有错误和警告|  

###10.3 异常处理
####10.3.1  实例4————一场的基本处理方法
异常处理用于在制定的错误（异常）情况发生时改变脚本的正常流程。这种情况称为异常。当异常被触发时，通常会发生一下动作。  
(1) 当前代码状态被保存。  
(2) 代码执行被切换到预定义的异常处理器函数。  
(3) 根据错误情况，处理器也许会从保存的代码状态重新开始执行代码，终止脚本执行，或从代码中另外的位置继续执行脚本。  
当异常被抛出时，其后的代码不会继续执行， PHP 会尝试查找匹配的  “catch” 代码块。 如果异常没有被捕获，而且又没有使用  set_exception_handler() 做相应处理的话，那么将发生一个严重的错误，并且输出  “Uncaught Exception” (未捕获异常)的错误消息。  
案例：  
//创建带有异常的函数
function checkNum($number){  
if($number>1){  
throw new Exception("Value must be 1 or below");  
}  
return true;  
}  
//抛出异常  
checkNum(2);  
***
//创建可抛出一个异常的函数  
function checkNum($number){  
if($number>1){  
throw new Exception("数值必须小于或等于1");  
}  
return true;  
}  
//在  "try" 代码块中触发异常  
try{  
checkNum(2);  
//如果没有异常，则会显示以下信息  
echo '没有任何异常';  
}  
//捕获异常  
catch(Exception $e){  
echo '异常信息：‘.$e->getMessage();  
}  
####10.3.2 实例5————自定义的异常处理器  
class customException extends Exception{  
public function errorMessage(){  
//错误消息  
$errorMsg='异常发生的行：'.$this->getLine().'in'.$this->getFile().':<b>'.$this->getMessage().'</b>不是一个有效的邮箱地址';  
return $errorMsg;  
}  
}  
$email ="someone@example.321com";  
try{  
//检查是否符合条件  
if(filter_var($email,FILTER_VALIDATE_EMAIL)===FALSE){  
//如果邮件地址无效，则抛出异常  
throw new customException($email);  
}  
}  
catch (customException $e){  
//显示自定义的消息  
echo $e->errorMessage();  
}  
####10.3.3 实例6————处理多个异常  
class customException extends Exception{  
public function errorMessage(){  
//定义错误信息  
$errorMsg='错误消息的行：'.$this->getLine().'in'.$this->getFile().':<b>'.$this->getMessage().'</b>不是一个有效的邮箱地址';  
return $errorMsg;  
}  
}  
$email = "someone@yahoo.com";  
try{  
//检查是否符合条件  
if(filter_var($email, FILTER_VALIDATE_EMAIL)===FALSE){  
//如果邮箱地址无效，则抛出异常  
throw new customException($email);  
}  
//检查邮箱是否为雅虎邮箱  
if(strpos($email,"yahoo")!==FALSE){  
throw new Exception("$email是一个雅虎邮箱");  
}  
}  
catch(customException $e){  
echo $e->errorMessage();  
}  
catch(Exception $e){  
echo $e->getMessage();  
}  
####10.3.4 实例7————设置顶层异常处理器  
案例：  
function myException($exception){  
echo "<b>异常时：</b>",$exception->getMessage();  
}  
set_exception_handler('myException');  
throw new Exception('正在处理未被捕获的异常');  
###10.4  实例8————处理异常或错误  
$DOCUMENT_ROOT=$_SERVER['DOCUMENT_ROOT'];  
@$fp=fopen("$DOCUMENT_ROOT/book.txt",'rb');  
try{  
if(!$fp){  
throw new Exception("文件路径有误或找不到文件。");  
}  
}catch(Exception $exception){  
echo $exception->getMessage();  
echo "在文件".$exception->getFile()."的".$excepton->getLine()."行。<br/>";  
}  
@fclose($fp);  


